core/tracestack.lua

```diff
@@ -1,147 +1,137 @@
+          info = info .. " near " .. _formatLocation(self[i], self[i].file == top.file)
           break
         end
       end
     end
     -- Print after, if it is in a relevant file
-    if afterFrame and (not base or afterFrame.file == base.file) then
-      info = info .. " after " .. afterFrame:toLocationString(--[[skipFile=]] true)
+    if self.lastPopped and (not base or self.lastPopped.file == base.file) then
+      info = info .. " after " .. _formatLocation(self.lastPopped, true)
     end
     return info
   end,
 
-  -- Push a command frame on to the stack to record the execution trace for debugging.
+  -- Push a command to the stack to record the execution trace for debugging.
   -- Carries information about the command call, not the command itself.
   -- Must be popped with `pop(returnOfPush)`.
-  pushCommand = function(self, tag, line, column, options, file)
+  pushCommand = function(self, tag, options, file, line, column)
     if not tag then
       SU.warn("Tag should be specified for SILE.traceStack:pushCommand", true)
     end
-    return self:_pushFrame({
-        tag = tag or "???",
-        file = file or SILE.currentlyProcessingFile,
-        line = line,
-        column = column,
-        options = options or {}
-      })
+    local file = file or SILE.currentlyProcessingFile
+    return self:_push(file, line, column, tag, options)
   end,
 
 
-  -- Push a command frame on to the stack to record the execution trace for debugging.
+  -- Push a command to the stack to record the execution trace for debugging.
   -- Command arguments are inferred from AST content, any item may be overridden.
   -- Must be popped with `pop(returnOfPush)`.
-  pushContent = function(self, content, tag, line, column, options, file)
+  pushContent = function(self, content, tag)
+    local tag = tag or content.tag
     if type(content) ~= "table" then
       SU.warn("Content parameter of SILE.traceStack:pushContent must be a table", true)
       content = {}
     end
-    tag = tag or content.tag
     if not tag then
       SU.warn("Tag should be specified or inferable for SILE.traceStack:pushContent", true)
     end
-    return self:_pushFrame({
-        tag = tag or "???",
-        file = file or content.file or SILE.currentlyProcessingFile,
-        line = line or content.line,
-        column = column or content.col,
-        options = options or content.attr or {}
-      })
+    local file = content.file or SILE.currentlyProcessingFile
+    return self:_push(file, content.line, content.col, tag, content.attr)
   end,
 
   -- Push a text that is going to get typeset on to the stack to record the execution trace for debugging.
   -- Must be popped with `pop(returnOfPush)`.
-  pushText = function(self, text, line, column, file)
-    return self:_pushFrame({
-        text = text,
-        file = file,
-        line = line,
-        column = column
-      })
+  pushText = function(self, text)
+    return self:_push(nil, nil, nil, nil, nil, text)
```

Comment:
It's not just verbosity. Lua tends to be verbose and often verbose is more readable, but the `pushX()` functions basically existed as syntax shortcuts to supply a well formed item on the stack. This included programming logic for fallback and default values. Because the programming logic was duplicated in several places my attempts to modify/extend it for tracking other things were failing. This simplification was not just to be less verbose, it was to consolidate the logic so there was one place to change it, not 3+.

Reply:
The thing is, the logic can be slightly different for each frame type, as different frames carry different contextual information. And especially for extending the system, it should be valuable that each frame can handle things differently, with different fields, etc.

The consolidation of all things into _push which takes fixed set of arguments basically means, that there can ever be only two types of frames: command frames (for which the `tag` field is useful) and text frames (for which `text` is useful). Anything else (custom payload frame?) is impossible, because it has to pass through the `_push`, which does not support it.

Additionally, this completely breaks `toLocationString` (or your `_formatLocation`), because that determines the type of the frame by looking at which fields are present. And since you always set `tag` to something in `_push`, it will consider all stack frames to be command frames and will never print any text. (Changing it to check `text` instead is just a poor workaround.)

The correct way to extend this to support different frame types would be to write different `pushSomething` functions. Sure, there would be a tiny bit of duplication, but it will be very easy to reason about and easy to maintain, because when one type of the frame changes (for example when `pushText` gets some additonal information, let's say the font or something) it will be trivial to just modify `pushText` without having to worry about what that means for `pushCommand`.

You argue from the different standpoint, from the point of changing something inside the system. However, my counterpoint is that if something internally changes so drastically, that it sends "ripples" outwards, it is necessary to check and potentially fix all affected places anyway. Keeping all logic in the same method for all possible cases will just create a maintainers nightmare, especially in Lua.